### ISSUE TITLE
Create Replica never returns if RunAsync is endless

### ISSUE THREAD

**Issue by sportfort** [loekd/ServiceFabric.Mocks#82]
In your Tests project, you have the following Test:

 ```
       [TestMethod]
        public async Task TestServiceState_InMemoryState_PromoteActiveSecondary()
        {
            var replicaSet = new MockStatefulServiceReplicaSet<MyStatefulService>(CreateStatefulService, CreateStateManagerReplica);
            await replicaSet.AddReplicaAsync(ReplicaRole.Primary, 1);
            await replicaSet.AddReplicaAsync(ReplicaRole.ActiveSecondary, 2);
            await replicaSet.AddReplicaAsync(ReplicaRole.ActiveSecondary, 3);

            const string stateName = "test";
            var payload = new Payload(StatePayload);

            //insert data
            await replicaSet.Primary.ServiceInstance.InsertAsync(stateName, payload);
            //promote one of the secondaries to primary
            await replicaSet.PromoteActiveSecondaryToPrimaryAsync(2);
            //get data
            var payloads = (await replicaSet.Primary.ServiceInstance.GetPayloadsAsync()).ToList();

            //data should match what was inserted against the primary
            Assert.IsTrue(payloads.Count == 1);
            Assert.IsTrue(payloads[0].Content == payload.Content);

            //the primary should not have any in-memory state
            var payloadsFromOldPrimary = await replicaSet[1].ServiceInstance.GetPayloadsAsync();
            Assert.IsTrue(!payloadsFromOldPrimary.Any());
        }

```

The problem is this test is not working, if service is executing endless loop in RunAsync, let's say for background task.

Some details:

AddReplicaAsync will wait for CreateAsync:
```
        public async Task AddReplicaAsync(ReplicaRole role, long? replicaId = null, int activationDelayMs = 0)
        {
            var serviceContext = MockStatefulServiceContextFactory.Create(CodePackageActivationContext, ServiceTypeName, ServiceUri, Guid.NewGuid(), replicaId ?? _random.Next());
            var stateManager = _stateManagerFactory(serviceContext, _reliableStates);
            var replica = new MockStatefulServiceReplica<TStatefulService>(_serviceFactory, serviceContext, stateManager);
            await replica.CreateAsync(role);
            _replicas.Add(replica);
        }
```

and then CreateAsync will wait for RunAsync which will never return

```
        public async Task CreateAsync(ReplicaRole role)
        {
            await OpenAsync(ReplicaOpenMode.New);

            if (role == ReplicaRole.Primary)
            {
                await OpenServiceReplicaListeners();
                await ChangeRoleAsync(role);
                await RunAsync();
            }
            else
                await ChangeRoleAsync(role);
        }
```

My suggestion is to wait for RunAnync to be completed in CreateAsync. Thank you.

**Comment by loekd** [comment]
I'm not sure I understand the problem. The code does wait for it to complete in the snippet, right?

**Comment by sportfort** [comment]
Yes, CreateAsync will never exit, in case of RunAsync is doing some background operations in an endless loop. This is the problem.

**Comment by Patrick-Spiegel** [comment]
@loekd I can confirm that this happens. E.g. if you keep the auto generated code from the sf template.

        protected override async Task RunAsync(CancellationToken cancellationToken)
        {
            long iterations = 0;

            while (true)
            {
                cancellationToken.ThrowIfCancellationRequested();

                ServiceEventSource.Current.ServiceMessage(this.Context, "Working-{0}", ++iterations);

                await Task.Delay(TimeSpan.FromSeconds(1), cancellationToken).ConfigureAwait(false);
            }
        }

**Comment by VaclavK** [comment]
@loekd also stumbled upon this, quite basic scenario - service with continuous dequeue in RunAsync and remoting listener open accepting message from outside and queuing them

I want to write  a test of switching promoting secondary to primary and queuing more message and seeing them being dequeued and acted up - which I can check via mock verifiable assertions

I looked briefly at the code and see that one - quick and dirty perhaps - option is to give caller access to task - so that the test can run it - e.g. till the assertion is met along with some max duration perhaps

loads of these loops will follow the pattern of service called (can be mocked) and some sleep

this would require swapping the order of replica added to the list and awaiting on the RunAsync - 

https://github.com/VaclavK/ServiceFabric.Mocks/commit/4fd3b98f5bf4cc7c1b2ecf006d62bc287a55b759

sample of my test code - quick and dirty
```
   clusterNotifierMock.Setup(i => i.DistributeToCluster(It.IsAny<InflightMessage>())).Returns(Task.CompletedTask).Verifiable();
           

        var replicaSet = new MockStatefulServiceReplicaSet<ISummonNoobsBackendService.ISummonNoobsBackendService>((sc,stateManagerReplica2) =>
        new ISummonNoobsBackendService.ISummonNoobsBackendService(sc, stateManagerReplica2, clusterNotifierMock.Object), CreateStateManagerReplica);

        var T = replicaSet.AddReplicaAsync(ReplicaRole.Primary, 1);
        T.Wait(TimeSpan.FromSeconds(2));
        
        await replicaSet.AddReplicaAsync(ReplicaRole.ActiveSecondary, 2);
        await replicaSet.AddReplicaAsync(ReplicaRole.ActiveSecondary, 3);

        await replicaSet.Primary.ServiceInstance.IngestMessage("blah", JsonMessage);
        T.Wait(TimeSpan.FromSeconds(2));
        clusterNotifierMock.Verify();
```

I would like to hear your thoughs...

**Comment by loekd** [comment]
I've taken a similar route and changed the call to `RunAsync` into an invokation without an await within `CreateAsync`. The task is awaited when the service is closed or deleted. The `CancellationToken` passed to `RunAsync` is cancelled, just as it would be in the real runtime.
Added a unit test to prove the functionality.
@sportfort , @Patrick-Spiegel , @VaclavK thanks for the assistance in this issue!
Please let me know if this is fixed in v3.4.10.

**Comment by VaclavK** [comment]
yeah the fix looks real good
